### PR TITLE
[WIP] [GR-28697] Add JDK Flight Recorder support for Java native images.

### PR DESCRIPTION
**Pull request by christianhaeubl**
This is a work in progress PR that adds some JFR infrastructure to Native Image.

### CODE REVIEW

**Review by machaval — APPROVED** [review]

File: substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEvents.java [hunk 1]
```diff
@@ -0,0 +1,53 @@
...
+ * questions.
+ */
+package com.oracle.svm.jfr;
+
+/**
+ * The event IDs depend on the metadata.xml and therefore vary between JDK versions.
+ */
+public enum JfrEvents {
+    // TODO: we need to abstract the JDK version in some way.
+    MetadataEvent(0),
```
Comment (aimiebell) [inline substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrEvents.java:32]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrManager.java [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+    public static class Options {
+        @Option(help = "Enable Java Flight Recorder.")//
+        public static final RuntimeOptionKey<Boolean> FlightRecorder = new RuntimeOptionKey<>(false);
+
+        @Option(help = "Start flight recording with options.")//
+        public static final RuntimeOptionKey<String> StartFlightRecording = new RuntimeOptionKey<>("");
+    }
+
+    private enum JfrStartArgument {
+        Name("name"),
```
Comment (aimiebell) [inline substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrManager.java:238]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Method names should start with a lower case letter (NM_METHOD_NAMING_CONVENTION)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-method-names-should-start-with-a-lower-case-letter-nm-method-naming-convention).
Methods should be verbs, in mixed case with the first letter lowercase, with the first letter of each internal word capitalized.

File: substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrManager.java [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+            periodicEventSetup();
+            initRecording();
+        }
+    }
+
+    void teardown() {
+        if (Options.FlightRecorder.getValue()) {
+            // Everything should already have been torn down by JVM.destroyJFR(), which is called in
+            // a shutdown hook.
+            assert !SubstrateJVM.isInitialized();
```
Comment (aimiebell) [inline substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrManager.java:77]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Use of identifier that is a keyword in later versions of Java (NM_FUTURE_KEYWORD_USED_AS_IDENTIFIER)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-use-of-identifier-that-is-a-keyword-in-later-versions-of-java-nm-future-keyword-used-as-identifier).
The identifier is a word that is reserved as a keyword in later versions of Java, and your code will need to be changed in order to compile it in later versions of Java.

File: substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrNativeEventWriter.java [hunk 4]
```diff
@@ -0,0 +1,356 @@
...
+    private static boolean ensureSize(JfrNativeEventWriterData data, int requested) {
+        assert requested > 0;
+        if (!isValid(data)) {
+            return false;
+        }
+
+        int totalRequested = requested + SIZE_SAFETY_CUSHION;
+        if (getAvailableSize(data).belowThan(totalRequested)) {
+            if (!accommodate(data, getUsedSize(data), totalRequested)) {
+                assert !isValid(data);
```
Comment (aimiebell) [inline substratevm/src/com.oracle.svm.jfr/src/com/oracle/svm/jfr/JfrNativeEventWriter.java:191]: I detect that this code is problematic. According to the [Bad practice (BAD_PRACTICE)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#bad-practice-bad-practice), [Nm: Use of identifier that is a keyword in later versions of Java (NM_FUTURE_KEYWORD_USED_AS_IDENTIFIER)](https://spotbugs.readthedocs.io/en/stable/bugDescriptions.html#nm-use-of-identifier-that-is-a-keyword-in-later-versions-of-java-nm-future-keyword-used-as-identifier).
The identifier is a word that is reserved as a keyword in later versions of Java, and your code will need to be changed in order to compile it in later versions of Java.

**Pull request merged**